src/tirith/core/core.py

```diff
@@ -69,15 +71,84 @@ def generate_evaluator_result(evaluator_obj, input_data, provider_module):
+    deleted_var_names: List[str] = []
+
+    class RemoveNoneConstantAndName(ast.NodeTransformer):
+        def visit_Constant(self, node: ast.Constant) -> Any:
+            if node.value is None:
+                return None
+            self.generic_visit(node)
+            return node
+
+        def visit_Name(self, node: ast.Name) -> Any:
+            deleted_var_names.append(node.id)
+            return None
+
+    class FixBoolOp(ast.NodeTransformer):
+        def visit_BoolOp(self, node: ast.BoolOp) -> Any:
+            if len(node.values) == 1:
+                # When there's only one child of a BoolOp, make that child
+                # as the parent so the tree becomes valid
+                return node.values[0]
+            if len(node.values) == 0:
+                # Consider if we need to return a True node here instead
+                return None
+            self.generic_visit(node)
+            return node
+
+    class FixUnaryOp(ast.NodeTransformer):
+        def visit_UnaryOp(self, node: ast.UnaryOp) -> Any:
+            if getattr(node, "operand", None) is None:
+                # When UnaryOp has no child, delete the node
+                return None
+            self.generic_visit(node)
+            return node
+
+    tree = ast.parse(eval_str, mode="eval")
+
+    compiled_code = None
+    tries_count = 0
+    is_tree_compilable = False
+    while not is_tree_compilable and tries_count <= MAX_TRIES:
+        try:
+            tries_count += 1
+            # Clean the tree from None (skip mark) and any variable names
+            tree = RemoveNoneConstantAndName().visit(tree)
+
+            if getattr(tree, "body", None) is None:
+                tree.body = ast.Constant(value=None)
+
+            ast.fix_missing_locations(tree)
+            compiled_code = compile(tree, "<string>", "eval")
+            is_tree_compilable = True
+        except ValueError as e:
+            logger.debug(e)
+            tree = FixBoolOp().visit(tree)
+        except TypeError as e:
+            logger.debug(e)
+            tree = FixUnaryOp().visit(tree)
+
+    return compiled_code, deleted_var_names
+
+
+def final_evaluator(eval_string: str, eval_id_values: Dict[str, Optional[bool]]) -> Tuple[bool, List[str]]:
     """
     Evaluate a given boolean expression string `eval_string` based on the boolean
     values provided by `eval_id_values`.
 
     Variable that has the value of `None` (we use it to mark a check as skipped) will
-    be replaced with `True`. This is due to the truthy value of None equals to False
+    be removed from the expression. This is due to the truthy value of None equals to False
     which will interfere with the final evaluation result.
 
+    All variables that are used within `eval_string` but not in `eval_id_values` will be
+    removed from the `eval_string` prior processing the `eval_string`.
+
     Example usage:
     >>> final_evaluator("(!(pol_check_1  &&  pol_check_2)  && pol_check_3 ) && pol_check_4", {
```

Comment:
I forgot to update the example usage in the docstring, will do a followup commit after this commit has been reviewed

Reply:
@refeed looks good. Please continue